python_modules/dagster/dagster/_core/definitions/dependency.py

```diff
@@ -486,22 +487,25 @@ def from_dict(cls, dict_repr: Mapping[str, Any]) -> "NodeHandle":
         return NodeHandle(name=dict_repr["name"], parent=parent)
 
 
+T_OptionalNodeHandle = TypeVar("T_OptionalNodeHandle", bound=Optional[NodeHandle])
+
+
 @dagster_model(checked=False)
-class NodeInputHandle:
+class NodeInputHandle(Generic[T_OptionalNodeHandle]):
     """A structured object to uniquely identify inputs in the potentially recursive graph structure."""
 
-    node_handle: NodeHandle
+    node_handle: T_OptionalNodeHandle
     input_name: str
 
     def __str__(self) -> str:
         return f"{self.node_handle}:{self.input_name}"
 
 
 @dagster_model(checked=False)
-class NodeOutputHandle:
+class NodeOutputHandle(Generic[T_OptionalNodeHandle]):
     """A structured object to uniquely identify outputs in the potentially recursive graph structure."""
 
-    node_handle: NodeHandle
+    node_handle: T_OptionalNodeHandle
     output_name: str
 
     def __str__(self) -> str:
```

Comment:
should we update `__str__` or do you want `None:some_input`

Reply:
I guess the alternative would be just `:some_input`?  `None` seems a little more explicit to me (although I guess can have collisions if a node is named "None"?). My instinct would be to leave it as is, but also happy to change it.

Reply:
> although I guess can have collisions if a node is named "None"

this is the strongest case to change it - I think these strings are just for debugging so i expect it doesnt matter much